core/common/src/main/java/alluxio/resource/SharedResourcePool.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.resource;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.PriorityQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class implements a SharedResourcePool. It manages resources under a different
+ * model than an exclusive {@link Pool}: in a pool, resources are owned exclusively,
+ * and the workflow is: user acquires a resource from the pool, owns it and works on it
+ * exclusively for a time, and then return it back to the pool so that others can acquire it.
+ * <p></p>
+ * A shared pool manages <b>Shareable Resources</b>, i.e., a list of resources shared among all
+ * users. When users acquire a resource from queue, they might get a new one, or they might get
+ * a used one to share with other users. The idea is that the underlying resources can be safely
+ * shared so that acquire never needs to wait or fail.
+ * <p></p>
+ * {@link SharedResourcePool} implements the shared pool model by maintaining a priority queue
+ * of resources ordered by current reference count.
+ *
+ * @param <T> The inner resource type
```

Comment:
The implicit but critical requirement on `T` is that it must be thread-safe *and* (effectively) stateless. Since these requirements cannot be expressed via Java's type system, we need to add more comments here.

Reply:
Maybe add some marker interface `ThreadSafe` and `Stateless` as a requirement for `T`, and implement them for the existing `T`s such as the several clients.